src/DelimitedDataParser/DelimitedDataReader.cs

```diff
@@ -304,7 +304,7 @@ public override IEnumerator GetEnumerator()
 
         public override Type GetFieldType(int ordinal)
         {
-            if (ordinal < 0 || ordinal > _currentRow.Count - 1)
+            if (!_yieldExistingRow && (ordinal < 0 || ordinal > _currentRow.Count - 1))
```

Comment:
As above. Should we not instead be modifying the checks that we perform depending on whether `_yieldExistingRow` is set? For example, if this is `false`, then we check against `_fieldNameLookup.Count`, otherwise we check against `_currentRow.Count`?